### PR TITLE
Feature/python

### PR DESCRIPTION
**Pull request by whaeck**
Starting python bindings for ACEtk - currently limited to the generic ACE table.

Updated ACEtk code with documentation, adding header guards, updated tests.

This is the first of a few upcoming pull requests.

### CODE REVIEW

**Comment by david8dixon** [comment]
I thought you said this was a big PR - only 72 files changed...

File: cmake/release_dependencies.cmake [hunk 1]
```diff
@@ -45,6 +45,11 @@ FetchContent_Declare( Log
     GIT_TAG         52962b7796afe37ef1d8f7edb4bf9ecb1b868d15
     )
 
+FetchContent_Declare( pybind11
+    GIT_REPOSITORY  https://github.com/pybind/pybind11
+    GIT_TAG         f1abf5d9159b805674197f6bc443592e631c9130 # tag: v2.6.1
+    )
+
 FetchContent_Declare( range-v3-adapter
```
Comment (david8dixon) [inline cmake/release_dependencies.cmake:53]: Is there still a reason for the adapter?

Comment (whaeck) [reply cmake/release_dependencies.cmake:53]: That's going to be another pull request (see #32)

**Comment by david8dixon** [comment]
Is there code coverage?

**Comment by david8dixon** [comment]
It doesn't appear to me that the bindings are being built...

**Comment by whaeck** [comment]
No, we switched that off some time ago. We used to get them through travis but our compile times became too high so we abandoned that. I did the coverage reports by hand for a while using gcov but since I was the only one actually doing that, I stopped doing that too. I should look into getting that back up and running though.

**Comment by whaeck** [comment]
There's a separate target named ACEtk.python that is excluded from the make all target.

**Comment by david8dixon** [comment]
ah ok I don't recall that being the case from previous experience with pybind11 but it has been a few years.

**Comment by david8dixon** [comment]
We should really consider removing the ACEtk.python target as a part of this PR. I feel that it would be much better to have it configurable. If I do the following:

```
make ACEtk.python
ctest -N
```

I get a bunch of noise because all the other test executables weren't built.

**Comment by nathangibson14** [comment]
How would you want the Python target to be handled?

Right now, we have it as a separate target, so that ACEtk could be included into a C++ application without the Python hoopla, and yes, the ctest is only set up for the C++ tests.  Our Python test suite is all in unittest, which can be run from a script or from a Python invocation.

Long term, the plan is to move away from building the Python via CMake (at least directly) and move to something like setuptools or pip, as that'd make installation of just the Python quite a bit smoother.

But all that said, we're only pretending we know what we're doing over here, so any advice on the build system is appreciated!

**Comment by david8dixon** [comment]
For now I would make the bindings a cmake option. Maybe have them off by default. Then the python build products and tests can be added conditionally depending on how one configures. This is a somewhat minor detail based on the stumbling around I did initially while trying to build and test.

**Comment by david8dixon** [comment]
Oh and I pushed up a change that has the python unit tests executed with the ctest driver.

**Comment by whaeck** [comment]
As far as I know, the C++ unit tests are already a CMake option that is switched on by default. If we add the python bindings as a CMake option that is also on by default, we can use that option to switch off the bindings in another library that uses ACEtk as a dependency - right?

We still need to do an overhaul of the build system scripts in devtools (one of the things that needs to be done is setting up the CMake python boilerplate automatically) so if you can come up with a proposed solution, that would be great.

Let's see if we can meet next week when we're all onsite to discuss/review a solution.

BTW: this would simplify the CI as well :-)

**Comment by david8dixon** [comment]
Why don’t I create  a branch with a prototype and we can go from there.

**Comment by whaeck** [comment]
@david8dixon Works for me ;-)

**Comment by david8dixon** [comment]
any idea why the checks aren't working?

**Comment by whaeck** [comment]
I've poked the CI to force a build and it seems to work. Did you see something in particular that went wrong?

**Review by david8dixon — APPROVED** [review]

**Comment by david8dixon** [comment]
I just wanted to be sure tests passed with the build system changes merged into this branch.

**Comment by david8dixon** [comment]
once this is in I will rebase the ranges branch and review

**Pull request merged**